Debugger part 3: display call stack and variable values

jfehrle opened the pull request:
@MSoegtropIMC I hope you'll try this version soon.  Please see the documentation in #14281.  In addition: Clicking on call stack entry (or Up/Down keys) jump to the associated location and shows variable values for that stack frame.  In the variables panel, click on the triangle to see lengthy values.  I recommend detaching the debugger panel for debugging multi-file scripts.

Some current limitations:
- Some call stack entries have no associated source code location.
- Still refining some of the stack entries, e.g. "??? LtacConstrInterp"
- Some variable values are not shown in a useful form, e.g. as "<genarg:tacvalue>"
- Can't copy variable values to the clipboard

Please create issues for any bugs you find and email me with your questions and comments.

Depends on and incorporates #14252

jfehrle commented:
@MSoegtropIMC Do you plan to try this sometime soon?  I've made several fixes and improvements, but it looks like I may need to rebase it before GitHub shows the build info on this page.

jfehrle commented:
@ppedrot Give me a couple of days to fix the broken tests and do some minor cleanup before you dive too deeply into reviewing this.

jfehrle commented:
@ppedrot Ready for review.

Zimmi48 commented:
If you intend to squash your commit before the PR is merged, then you can safely ignore the linter failure. It is failing because some intermediate commit fails to build (with warnings as errors).

Zimmi48 commented on doc/changelog/09-coqide/14644-debugger_stack.rst:
```diff
@@ -0,0 +1,8 @@
+- **Added:**
+  Adds a preliminary visual debugger.  Supports setting breakpoints
```
No need to repeat "add":
```suggestion
  A preliminary visual debugger.  Supports setting breakpoints
```

jfehrle commented:
Yes, I expect to squash before merging.

jfehrle commented:
Ali, I appreciate your comments and suggestions, but I'd like to keep the PR focused on the code review.  Please put your thoughts in issues rather than here (where they are likely to be lost).

> Can we customise this or at least set the default to something less blue? It's currently very similar to highlighting text.

See Edit / Preferences / Colors.  The setting will be persistent.

> Might it also be a good idea to have a help menu in the debug drop down, especially hinting that you need to do Set Ltac Debug or equivalent.

Not sure what you have in mind.  A link to the manual?  (open a new issue, don't reply here)

maximedenes commented:
> My request is that refs should not be exposed _in the public API_ in `debugHook`, but I think it is fine and even reasonable to have tactic_debug store the debugger state in a ref [privately]

That's literaly what was said during the Coq call, AFAICT :)

jfehrle commented:
>> have tactic_debug store the debugger state in a ref [privately]
> That's literaly what was said during the Coq call, AFAICT :)

I don't recall hearing that in the meeting.  In the subsequent discussion here, @ppedrot wanted to put the breakpoints in `TacStore`.  Hopefully he will not object to storing it in a ref in `tactic_debug`.

jfehrle commented:
> No. The format exported to UI's should be the same regardless of the backend involved. On the contrary, they should not export anything, but you may need both debuggers to communicate at their gateway point, that's not to be exposed on debugHook

OK, we'll perhaps continue discussion on this point in a future PR.

jfehrle commented on toplevel/coqtop.ml:
```diff
@@ -141,6 +141,7 @@ let ltac_debug_answer = let open DebugHook.Answer in function
       Format.fprintf !Topfmt.err_ft "@[%a@]@\n%!" Pp.pp_with o
     | Init ->
       Format.fprintf !Topfmt.err_ft "@[%a@]@\n%!" Pp.pp_with (str "Init")
+    | Vars _ -> ()  (* todo *)
```
`Vars` has a value of type `(string * Pp.t) list`.  Is there a simple way to marshall this using the existing type definition for the `Db_vars` message in `Xmlprotocol`?  If I need to write new marshalling code, what should that look like?

ejgallego replied:
Should be just `of_pair of_string of_pp` ? Actually this marshalling code is super annoying, and hindering your work big time.

It wouldn't take long to remove xmlprotocol at all and have `ppx_deriving_yojson` generate the file automatically; tho users of the raw protocol would have to adapt.

jfehrle replied:
Close, should be `let of_vars vars = of_list (of_pair of_string of_pp) vars`.  But, as usual, it took a fair bit of time to reverse engineer how to do this.

I want to keep xmlprotocol for this PR, but I'm open to considering changes a bit later.  There are a bunch of things in CoqIDE that would benefit from simplification and refactoring.

jfehrle replied:
However the conversion belongs elsewhere; this particular line of code is for `coqtop`.  This should just be a `failwith`.  I'm not planning to add stack and variable display to the coqtop version.  Indeed, I'd like to drop the support in `coqtop` since it's so much less usable.  But not today.

jfehrle reviewed (commented):
@ejgallego What's the easiest way to handle the marshalling for Answer?  See the last commit.

ejgallego commented on ide/coqide/idetop.ml:
```diff
@@ -671,15 +672,15 @@ let loop ( { Coqtop.run_mode; color_mode },_) ~opts:_ state =
-      | Goal msg -> "goal", (str "Goal:" ++ fnl () ++ msg)
-      | Output msg -> "output", msg
-      | Init -> "init", str ""
+    let open DebugHook.Answer in
+    let open Xmlprotocol in
+    match ans with
+      | Prompt msg -> print_xml xml_oc (of_ltac_debug_answer ~tag:"prompt" msg)
+      | Goal msg -> print_xml xml_oc (of_ltac_debug_answer ~tag:"goal" (str "Goal:" ++ fnl () ++ msg))
+      | Output msg -> print_xml xml_oc (of_ltac_debug_answer ~tag:"output" msg)
+      | Init -> print_xml xml_oc (of_ltac_debug_answer ~tag:"init" (str ""))
+      | Vars v -> print_xml xml_oc (of_answer Ppcmds Db_vars v)
```
Note you want to put this code in xmlprotocol I think

jfehrle commented:
@ejgallego Updated.  I changed the set breakpoint, get stack and get vars messages to go through DebugHook.Action and DebugHook.Answer as you asked.  This took significant time; I got stuck a few times.  I hope we're done now except for squashing.

jfehrle commented:
ping @ejgallego

ejgallego commented:
> ping @ejgallego

Sorry I missed this due to the terrible github interface and bug (*), will have a look ASAP.

(*) I basically click on the PR updated notification, but I land on the buggy "changes" pages, so I need to get the full notification away :S

ejgallego reviewed (commented):
Looks much better, thanks @jfehrle ; couple of additional things:
- please document `debugHook.mli` , so I can be sure I understand the protocol
- why do we need to perform `set_break` in idetop.ml ? That deserves a comment too.

jfehrle commented:
Updated and squashed.  I added the protocol documentation and moved set_break into tactic_debug, although I'll probably need to reverse the latter when we support additional interpreters such as Ltac2.  I believe I fixed the coqtail overlay, so this should be good to go once the tests pass, which I believe they will.

jfehrle commented:
The two failing jobs appear unrelated.

ejgallego commented on toplevel/coqtop.ml:
```diff
@@ -139,12 +139,16 @@ let ltac_debug_answer = let open DebugHook.Answer in function
+    | Init ->
+      Format.fprintf !Topfmt.err_ft "@[%a@]@\n%!" Pp.pp_with (str "Init")
+    | Stack _ -> failwith "Stack" (* not supported in coqtop *)
+    | Vars _ -> failwith "Vars" (* not supported in coqtop *)
```
Why is this not supported? You could print it to the terminal? I'd suggest using a proper error message, not an anomaly.

jfehrle replied:
There's currently no code path that can generate these replies in coqtop.  The command line debugger is a candidate for removal--its usability is so much less than the visual debugger that it's probably not worth keeping around at all, let alone enhancing it.

SkySkimmer replied:
you should assert false or anomaly for things that are never supposed to execute, not failwith

jfehrle replied:
What do you have against `failwith`?

I changed it to `anomaly`.  I don't like `assert false` because it doesn't return a message string (unlike `failwith`).  At best you get a file and a line number in a stack trace, which you then need to look up in the exactly matching source code to learn anything, so harder for both developers and users.

ejgallego reviewed (approved):
Protocol parts look OK to me for now.

Changes in IDE are very extensive tho, including the modification of some tricky stuff regarding GTK and window sizing, etc...

Thus this needs approval of @coq/ltac-maintainers and @coq/coqide-maintainers too.

ejgallego reviewed (changes requested):
Oh, `dev/doc/changes.md` need updating [tho I'd prefer to keep the xml protocol changelog in the file itself, and I remember adding it to that effect, but seems it was removed at some point]

jfehrle commented:
> Thus this needs approval of @coq/ltac-maintainers and @coq/coqide-maintainers too.

@ppedrot would you finish the review and be the assignee?  IIUC, in our Zoom conversation a couple weeks ago you had said the IDE changes were OK.  Let's try to get this completed for 8.15.  Thanks.

jfehrle commented:
@ejgallego I updated `dev/doc/changes.md`.  Please take a look.

ejgallego commented:
> @ejgallego I updated `dev/doc/changes.md`. Please take a look.

Thanks. The point critical to me is the modification of the `Add` call that breaks compat with virtually all users. Maybe we could be a bit more explicit on what the changes are in the changelog so people know how to adapt their uses?

[There is also a more hacky approach to this kind of changes, which is to introduce a new call `add_with_loc` , so the old call is still preserved, at least for a while, similar to deprecation status]

ejgallego commented on dev/doc/changes.md:
```diff
@@ -1,5 +1,17 @@
+  to the "add" message) and added support for these primarily in tactic_debug.ml
+- Currently only for Ltac in CoqIDE, but readily extendable to other
+  scripting languages and IDEs.
+
```
I think items 1 and 3 may not be relevant in this file, as they are more like user facing changes? [And already in the main changelog I guess]

I would maybe remove them, and update item 2 to specify the relevant change to add in more detail. Something such as: "before this change, add returned `<xml> .... </xml>` , now, new attributes `<foo> ...` and `<bar>` have been added.

ejgallego replied:
Or provide a pointer to the document where the new return format is spec'ed.

jfehrle replied:
Hard to know what should and shouldn't appear in this file.  I removed items 1 and 3.  I added a pointer to `xmlprotocol.md`, which has detailed descriptions of all the messages and responses.  Note that "add" got extra parameters rather than extra return values.

jfehrle replied:
Note that this PR includes an overlay for coqtail to address this very point.  Aside from the PG prototype, are you aware of any other packages that use the xml protocol?

ejgallego replied:
There are some other users indeed, not very actively maintained, but given what we wrote on the protocol file we need to signal the protocol changed [or introduce backwards compatible changes]

Personally I'd keep the changelog for the protocol in the .mli file, in a format `-date-protocol: changes`

ejgallego replied:
Another good practice for this kind of changelogs is to signal explicitly that the change is backwards incompatible, thus you can write:
```
# XML protocol (IMHO better under its own header)
- The add message will now return extra fields for location of the added sentence (warning, backwards-incompatible change, see protocol.md file for more details)
- 4 new "db_*" messages have been added to support the ltac_debugger
```

ejgallego commented on ide/coqide/session.ml:
```diff
@@ -506,27 +561,22 @@ let build_layout (sn:session) =
-      eval_paned#set_position estimated_vertical_handle_position;
-    if estimated_horizontal_handle_position > 0 then
-      state_paned#set_position estimated_horizontal_handle_position in
+
```
This is the main change in the coqide codebase that worries me, wasn't this submitted as a different PR?

I don't recall the outcome of the discussion, but I'm wary folding this kind of unrelated modifications in this PR.

cc: @coq/coqide-maintainers

jfehrle replied:
I reversed this.  We earlier decided not to make the change in the PR, #14536.

ejgallego commented:
> I reversed this. We earlier decided not to make the change in #14536.

Good thanks. @SkySkimmer I added you to the review as not to have a mess with the 8.15 branch; I'd like @ppedrot to give a peek at this if possible.

ppedrot reviewed (approved):
Go ahead.

ejgallego commented:
@coqbot: merge now

@jfehrle you may consider tweanking the dev/doc/changes.md file if you want.

jfehrle commented:
@ejgallego @ppedrot Thanks for the review!  Glad we got it into 8.15 just under the wire.

mattam82 commented:
It seems this might have broken VSCoq on master due to the changes in the XML protocol

jfehrle commented:
Would you open an issue for this and provide more details, e.g.:
- "starting interpretation" meaning trying to execute a tactic or command?
- What are the effects (other than the messages)?
- Were you trying to use the debugger or had a `Set Ltac Debug.` in your script?
- What version of vsCoq/when built?

The debugger changed the parameters and reply for the "add tactic/command" message.
That should be an easy fix in VSCoq if that is the cause.  However, my impression was that
VSCoq was not using the XML protocol.  (8.13 also modified that same message but I never
heard about any issue with VSCoq.)

Attn: VSCoq maintainers @maximedenes @fakusb

Blaisorblade commented:
> However, my impression was that
> VSCoq was not using the XML protocol.

It certainly is using it — that's how it supports async proofs!

jfehrle commented:
Please put your comments on the vsCoq issue here: #15282.
